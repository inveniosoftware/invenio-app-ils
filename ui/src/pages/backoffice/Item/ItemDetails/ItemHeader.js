import { toShortDate } from '@api/date';
import { CopyButton, DocumentAuthors } from '@components';
import { DocumentTitle } from '@components/Document';
import {
  DetailsHeader,
  DocumentIcon,
  ItemIcon,
} from '@pages/backoffice/components';
import { BackOfficeRoutes } from '@routes/urls';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { Header } from 'semantic-ui-react';

export class ItemHeader extends Component {
  render() {
    const { data } = this.props;
    const recordInfo = (
      <>
        <label className="muted">Physical copy</label> {data.metadata.pid}{' '}
        <CopyButton text={data.metadata.pid} />
        <br />
        <label className="muted">Created on</label> {toShortDate(data.created)}
        <br />
        <Link
          to={BackOfficeRoutes.documentDetailsFor(data.metadata.document_pid)}
        >
          see document <DocumentIcon />
        </Link>
      </>
    );
    return (
      <DetailsHeader
        title={
          <>
            <Header.Subheader>Medium: {data.metadata.medium}</Header.Subheader>
            {data.metadata.barcode}:{' '}
            <DocumentTitle
              document={data.metadata.document}
              short={true}
              truncate={true}
            />
          </>
        }
        subTitle={
          <>
            <DocumentAuthors metadata={data.metadata.document} prefix={'by '} />
          </>
        }
        pid={data.metadata.pid}
        icon={<ItemIcon />}
        recordType="Document"
        recordInfo={recordInfo}
      />
    );
  }
}

ItemHeader.propTypes = {
  data: PropTypes.object.isRequired,
};
