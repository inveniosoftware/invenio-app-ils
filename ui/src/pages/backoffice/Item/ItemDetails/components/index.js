export { ItemMetadata } from './ItemMetadata';
export { ItemPastLoans } from './ItemPastLoans';
export { ItemCirculation } from './ItemCirculation';
export { ItemActionMenu } from './ItemActionMenu';
