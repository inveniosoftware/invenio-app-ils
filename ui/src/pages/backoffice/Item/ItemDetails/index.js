import { connect } from 'react-redux';

import { deleteItem, fetchItemDetails } from './state/actions';
import ItemDetailsComponent from './ItemDetails';

const mapStateToProps = state => ({
  isLoading: state.itemDetails.isLoading,
  error: state.itemDetails.error,
});

const mapDispatchToProps = dispatch => ({
  fetchItemDetails: itemPid => dispatch(fetchItemDetails(itemPid)),
  deleteItem: itemPid => dispatch(deleteItem(itemPid)),
});

export const ItemDetails = connect(
  mapStateToProps,
  mapDispatchToProps
)(ItemDetailsComponent);
