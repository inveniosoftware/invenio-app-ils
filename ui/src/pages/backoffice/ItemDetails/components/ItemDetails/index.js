import { connect } from 'react-redux';

import ItemDetailsComponent from './ItemDetails';

const mapStateToProps = state => ({
  isLoading: state.itemDetails.isLoading,
  error: state.itemDetails.error,
  hasError: state.itemDetails.hasError,
});

export const ItemDetails = connect(mapStateToProps)(ItemDetailsComponent);
