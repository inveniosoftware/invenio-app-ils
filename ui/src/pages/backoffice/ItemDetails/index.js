import { connect } from 'react-redux';

import { deleteItem, fetchItemDetails } from './state/actions';
import ItemDetailsContainerComponent from './ItemDetailsContainer';

const mapDispatchToProps = dispatch => ({
  fetchItemDetails: itemPid => dispatch(fetchItemDetails(itemPid)),
  deleteItem: itemPid => dispatch(deleteItem(itemPid)),
});

export const ItemDetailsContainer = connect(
  null,
  mapDispatchToProps
)(ItemDetailsContainerComponent);
