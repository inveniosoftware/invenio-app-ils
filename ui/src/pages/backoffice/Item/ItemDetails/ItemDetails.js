import { ItemActionMenu, ItemCirculation } from './components';
import { ItemHeader } from './ItemHeader';
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Container, Divider, Grid, Ref, Sticky } from 'semantic-ui-react';
import { Loader, Error } from '@components';
import { ItemMetadata, ItemPastLoans } from './components';

export default class ItemDetails extends Component {
  constructor(props) {
    super(props);
    this.menuRef = React.createRef();
    this.headerRef = React.createRef();
  }

  componentDidMount() {
    this.unlisten = this.props.history.listen(location => {
      if (location.state && location.state.itemPid) {
        this.props.fetchItemDetails(location.state.itemPid);
      }
    });
    this.props.fetchItemDetails(this.props.match.params.itemPid);
  }

  componentWillUnmount() {
    this.unlisten();
  }

  render() {
    const { isLoading, error, data } = this.props;
    return (
      <div ref={this.headerRef}>
        <Container fluid>
          <Loader isLoading={isLoading}>
            <Error error={error}>
              <Sticky context={this.headerRef} className="solid-background">
                <Container fluid className="spaced">
                  <ItemHeader data={data} />
                </Container>
                <Divider />
              </Sticky>
              <Container fluid>
                <Ref innerRef={this.menuRef}>
                  <Grid columns={2}>
                    <Grid.Column width={13}>
                      <Container className="spaced">
                        <Container className="spaced">
                          <ItemCirculation />
                          <ItemMetadata />
                          <ItemPastLoans />
                        </Container>
                      </Container>
                    </Grid.Column>
                    <Grid.Column width={3}>
                      <Sticky context={this.menuRef} offset={180}>
                        <ItemActionMenu offset={-180} />
                      </Sticky>
                    </Grid.Column>
                  </Grid>
                </Ref>
              </Container>
            </Error>
          </Loader>
        </Container>
      </div>
    );
  }
}

ItemDetails.propTypes = {
  fetchItemDetails: PropTypes.func.isRequired,
  deleteItem: PropTypes.func.isRequired,
};
