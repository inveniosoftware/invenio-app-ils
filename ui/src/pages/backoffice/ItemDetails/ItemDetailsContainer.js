import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Container } from 'semantic-ui-react';
import { ItemDetails } from './components';

export default class ItemDetailsContainer extends Component {
  constructor(props) {
    super(props);
    this.fetchItemDetails = this.props.fetchItemDetails;
    this.deleteItem = this.props.deleteItem;
  }

  componentDidMount() {
    this.unlisten = this.props.history.listen(location => {
      if (location.state && location.state.itemPid) {
        this.fetchItemDetails(location.state.itemPid);
      }
    });
    this.fetchItemDetails(this.props.match.params.itemPid);
  }

  componentWillUnmount() {
    this.unlisten();
  }

  render() {
    return (
      <Container>
        <ItemDetails />
      </Container>
    );
  }
}

ItemDetailsContainer.propTypes = {
  fetchItemDetails: PropTypes.func.isRequired,
  deleteItem: PropTypes.func.isRequired,
};
