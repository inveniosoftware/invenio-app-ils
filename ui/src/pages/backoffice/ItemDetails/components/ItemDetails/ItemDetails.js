import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Loader, Error } from '../../../../../common/components';
import { ItemMetadata, ItemPastLoans } from '../';

export default class ItemDetails extends Component {
  render() {
    const { isLoading, error } = this.props;
    return (
      <Loader isLoading={isLoading}>
        <Error error={error}>
          <ItemMetadata />
          <ItemPastLoans />
        </Error>
      </Loader>
    );
  }
}

ItemDetails.propTypes = {
  isLoading: PropTypes.bool.isRequired,
  hasError: PropTypes.bool.isRequired,
};
