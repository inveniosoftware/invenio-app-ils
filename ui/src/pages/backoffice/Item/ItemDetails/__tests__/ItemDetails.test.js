import testData from '@testData/items.json';
import { mount, shallow } from 'enzyme';
import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import ItemDetails from '../ItemDetails';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

jest.mock('../components/', () => {
  return {
    ItemMetadata: () => null,
    ItemCirculation: () => null,
    ItemPastLoans: () => null,
    ItemActionMenu: () => null,
  };
});

jest.mock('../../ItemDetails/', () => {
  return {
    ItemHeader: () => null,
  };
});

const data = {
  hits: {
    total: 2,
    hits: [{ metadata: testData[0] }, { metadata: testData[1] }],
  },
};

let store;
beforeEach(() => {
  store = mockStore({
    isLoading: false,
    data: data,
    hasError: false,
    itemDetails: data,
  });
  store.clearActions();
});

describe('ItemDetails tests', () => {
  let component;
  afterEach(() => {
    if (component) {
      component.unmount();
    }
  });

  const routerHistory = {
    listen: () => () => {},
  };
  const routerUrlParams = {
    params: {
      itemPid: 111,
    },
  };

  it('should load the details component', () => {
    const component = shallow(
      <ItemDetails
        history={routerHistory}
        match={routerUrlParams}
        fetchItemDetails={() => {}}
        deleteItem={() => {}}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it('should fetch item details on mount', () => {
    const mockedFetchItemDetails = jest.fn();
    component = mount(
      <BrowserRouter>
        <Provider store={store}>
          <ItemDetails
            history={routerHistory}
            match={routerUrlParams}
            fetchItemDetails={mockedFetchItemDetails}
            data={{ metadata: testData[0] }}
            deleteItem={() => {}}
          />
        </Provider>
      </BrowserRouter>
    );
    expect(mockedFetchItemDetails).toHaveBeenCalledWith(
      routerUrlParams.params.itemPid
    );
  });
});
