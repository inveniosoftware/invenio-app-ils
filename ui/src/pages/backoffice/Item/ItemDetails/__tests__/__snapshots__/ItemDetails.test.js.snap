// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemDetails tests should load the details component 1`] = `
<div>
  <Container
    fluid={true}
  >
    <Loader>
      <Error>
        <Sticky
          active={true}
          bottomOffset={0}
          className="solid-background"
          context={
            Object {
              "current": null,
            }
          }
          offset={0}
          scrollContext={[Window]}
        >
          <Container
            className="spaced"
            fluid={true}
          >
            <ItemHeader />
          </Container>
          <Divider />
        </Sticky>
        <Container
          fluid={true}
        >
          <Ref
            innerRef={
              Object {
                "current": null,
              }
            }
          >
            <Grid
              columns={2}
            >
              <GridColumn
                width={13}
              >
                <Container
                  className="spaced"
                >
                  <Container
                    className="spaced"
                  >
                    <ItemCirculation />
                    <ItemMetadata />
                    <ItemPastLoans />
                  </Container>
                </Container>
              </GridColumn>
              <GridColumn
                width={3}
              >
                <Sticky
                  active={true}
                  bottomOffset={0}
                  context={
                    Object {
                      "current": null,
                    }
                  }
                  offset={180}
                  scrollContext={[Window]}
                >
                  <ItemActionMenu
                    offset={-180}
                  />
                </Sticky>
              </GridColumn>
            </Grid>
          </Ref>
        </Container>
      </Error>
    </Loader>
  </Container>
</div>
`;
